src/python/pants/backend/python/typecheck/mypy/rules.py

```diff
@@ -142,59 +182,48 @@ async def mypy_typecheck(
-    #     has code that requires Python 3.8+, and if so, use a tighter requirement.
-    #
-    #     On top of this, MyPy parses the AST using the value from `python_version` from mypy.ini.
-    #     If this is not configured, it defaults to the interpreter being used. This means that
-    #     running MyPy with Py35 would choke on f-strings in Python 3.6, unless the user set
-    #     `python_version`. We don't want to make the user set this up. (If they do, MyPy will use
-    #     `python_version`, rather than defaulting to the executing interpreter).
-    #
-    #  * When resolving third-party requirements, we should use the actual requirements. Normally,
-    #     we would merge the requirements.pex with mypy.pex via `--pex-path`. However, this will
-    #     cause a runtime error if the interpreter constraints are different between the PEXes and
-    #     they have incompatible wheels.
-    #
-    #     Instead, we teach MyPy about the requirements by extracting the distributions from
-    #     requirements.pex and setting EXTRACTED_WHEELS, which our custom launcher script then
-    #     looks for.
-    code_interpreter_constraints = PexInterpreterConstraints.create_from_compatibility_fields(
-        (
-            tgt[PythonInterpreterCompatibility]
-            for tgt in typechecked_transitive_targets.closure
-            if tgt.has_field(PythonInterpreterCompatibility)
-        ),
-        python_setup,
+    # If the user did not set `--python-version` already, we set it ourselves based on their code's
+    # interpreter constraints. This determines what AST is used by MyPy.
+    python_version = (
+        None
+        if partition.python_version_already_configured
+        else partition.interpreter_constraints.minimum_python_version()
     )
 
-    if not mypy.options.is_default("interpreter_constraints"):
-        tool_interpreter_constraints = mypy.interpreter_constraints
-    elif code_interpreter_constraints.requires_python38_or_newer():
-        tool_interpreter_constraints = ("CPython>=3.8",)
-    elif code_interpreter_constraints.requires_python37_or_newer():
-        tool_interpreter_constraints = ("CPython>=3.7",)
-    elif code_interpreter_constraints.requires_python36_or_newer():
-        tool_interpreter_constraints = ("CPython>=3.6",)
-    else:
-        tool_interpreter_constraints = mypy.interpreter_constraints
+    # MyPy requires 3.5+ to run, but uses the typed-ast library to work with 2.7, 3.4, 3.5, 3.6,
+    # and 3.7. However, typed-ast does not understand 3.8, so instead we must run MyPy with
+    # Python 3.8 when relevant. We only do this if if <3.8 can't be used, as we don't want a
+    # loose requirement like `>=3.6` to result in requiring Python 3.8, which would error if
+    # 3.8 is not installed on the machine.
+    tool_interpreter_constraints = PexInterpreterConstraints(
+        ("CPython>=3.8",)
```

Comment:
What if a user wants to specify 3.9 (3.10 is in the CPython git at this point)?

Reply:
Not just a trivial question as I very-very-occasionally-and-not-frequently test against that with pants to try out random things.

Reply:
They can! Via `--mypy-interpreter-constraints`. This is the default.

Or do you mean that we should make this more robust that we also check if it requires Py39 and we set automatically to >=3.9 if relevant? That'd be good to do, for sure. Let's save it for a followup, though, as we need to update Black too and I'm not even yet sure these tools themselves work with Py39. But ik the release is coming soon, and it would be good for us to try things out.

Reply:
Oh no! I'm totally fine if this can be set by hand.